dirdiff

```diff
@@ -0,0 +1,275 @@
+        "--debug",
+        action="store_true",
+        help="enable debug logging information",
+    )
+    parser.add_argument(
+        "--ignore-modified",
+        action="store_true",
+        help="don't show files that were modified",
+    )
+    parser.add_argument(
+        "--ignore-moved",
+        action="store_true",
+        help="don't show files that were moved around",
+    )
+    parser.add_argument(
+        "--ignore-additional",
+        action="store_true",
+        help="don't show files if they weren't either moved or modified",
+    )
+    parser.add_argument(
+        "--json",
+        action="store_true",
+        help="display output as JSON",
+    )
+    parser.add_argument(
+        "dirpath",
+        type=Path,
+        nargs="+",
+        help="dir path",
+    )
+    args = parser.parse_args()
+    if args.debug:
+        LOG.setLevel(logging.DEBUG)
+    return args
+
+
+@contextmanager
+def chdir(dirname):
+    oldpwd = Path().absolute()
+    try:
+        os.chdir(dirname)
+        yield
+    finally:
+        os.chdir(oldpwd)
+
+
+def apply_to_dir(dirpath, func):
+    return [
+        func(os.path.join(dirp, fname))
+        for dirp, _, filenames in os.walk(dirpath)
+        for fname in filenames
+    ]
+
+
+def file_hexdigest(filepath):
+    hasher = hashlib.sha1()
+    BUFSIZ = 65536
+    try:
+        with open(filepath, "rb") as f:
+            buf = f.read(BUFSIZ)
+            while len(buf) > 0:
+                hasher.update(buf)
+                buf = f.read(BUFSIZ)
+    except FileNotFoundError as e:
+        # if this is a symlink
+        if stat.S_ISLNK(os.lstat(filepath).st_mode):
+            hasher.update(b"")
+        else:
+            raise
+    return hasher.hexdigest()
+
+
+def preprocess_into_datastructs(rootdirs: List[Path]):
+    # let's collect all the files from all the rootdirs and build our base data structures
+    hashdiff_data: Dict[HexDigest, Dict[Path, List[DirDiffFile]]] = defaultdict(
+        lambda: defaultdict(list)
+    )
+    pathdiff_data: Dict[Path, List[DirDiffFile]] = defaultdict(list)
+
+    for rootdir in rootdirs:
+        rootdir_path = Path(rootdir)
+        with chdir(rootdir_path):
+            filelist = apply_to_dir(".", lambda x: Path(x))
+        for f in filelist:
+            try:
+                ddf = DirDiffFile(rootdir_path, f)
+                hashdiff_data[ddf.hexdigest][ddf.position].append(ddf)
+                pathdiff_data[ddf.position].append(ddf)
+            except PermissionError as e:
+                LOG.warning(f"error processing {f}: {e}")
+
+    return hashdiff_data, pathdiff_data
+
+
+def partition_matched_unmatched(
+    hashdiff_data: Dict[HexDigest, Dict[Path, List[DirDiffFile]]], rootdirs: List[Path]
+):
+    matched: Dict[HexDigest, List[DirDiffFile]] = defaultdict(list)
+    unmatched: Dict[HexDigest, List[DirDiffFile]] = defaultdict(list)
+
+    # dir1/foopath1/barfile1 is matched with dir2/foopath2/barfile2 if:
```

Comment:
following this comment, it seems that if
- foopath1 == foopath2
- hash(barfile1) == hash(barfile2).
- BUT barfile1 != barfile2 (file renamed IN PLACE), they are still considered matched?

Reply:
Yes, they are!

However, it's important to note that this comment has more of an explanatory purpose than a description of the algorithm. The code also considers the file names themselves. See (here)[https://github.com/jrabinow/homedir-scripts/blob/ea0f7ae7140ccd6df5c2774860cb9179a70869f3/dirdiff#L163] for matching and [here](https://github.com/jrabinow/homedir-scripts/blob/ea0f7ae7140ccd6df5c2774860cb9179a70869f3/dirdiff#L45) for path

I agree though, the comment could have explained things better. Updated.